Adding etcdctl install on demand script to make usage easier

mimiteto opened the pull request:
**What this PR does / why we need it**:
This PR should make a bit easier for persons who troubleshoot to install `etcdctl`

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
I mostly copied/adapted things based on the existing install script and the same script for wireguard.
To check if the modification works as expected I did a local build of the image and tried installing and using `etdctl`.
I am not sure if it's better to leave the current install script as well or leave it as-is.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```bugfix operator
`etcdctl` command is now available within the `ops-toolbelt` image. Upon first invocation it will be installed and will be fully usable.
```

petersutter commented on dotfiles/.install_on_demand/.etcdctl:
```diff
@@ -0,0 +1,49 @@
+    local _confirmation=""
+    while [[ ! $_confirmation =~ ^[$yes$no]$ ]]; do
+          read -p "${app} not found. Do you want to install it now? (Y/n) " -r _confirmation
+          if [[ $_confirmation =~ ^[$no]$ ]]; then
```
```bash
                     _                            _          _ _
  __ _  __ _ _ __ __| | ___ _ __   ___ _ __   ___| |__   ___| | |
 / _` |/ _` | '__/ _` |/ _ \ '_ \ / _ \ '__| / __| '_ \ / _ \ | |
| (_| | (_| | | | (_| |  __/ | | |  __/ |    \__ \ | | |  __/ | |
 \__, |\__,_|_|  \__,_|\___|_| |_|\___|_|    |___/_| |_|\___|_|_|
 |___/

Run ghelp to get information about installed tools and packages

root at 3438ec5ec3ad in /
$ etcdctl
etcdctl not found. Do you want to install it now? (Y/n)
etcdctl not found. Do you want to install it now? (Y/n)
etcdctl not found. Do you want to install it now? (Y/n)
```

Y is the default (as indicated by its uppercase), but hitting enter will not select the default option, which is ‘yes’.

mimiteto replied:
I copied this from `wg`.
Will adapt accordingly.

petersutter commented on dotfiles/.install_on_demand/.etcdctl:
```diff
@@ -0,0 +1,49 @@
+    rm -r "etcd-${version}-linux-amd64"
+
+
+  echo "You can now start using etcdctl. Just execute \"etcdctl\" to use it. See https://etcd.io/docs/v3.4/dev-guide/interacting_v3/ for more details."
+  echo "This tool assumes that it is being run in an ephemeral container in a pod. If this is not the case then please ensure that you provide the correct ecrtificates, the correct endpoint and have all necessary accesses."
+  echo "Certificates to be passed to the command should be mounted onto any container in the pod having a shared process namespace. Please run \"ps -A\" to list all processes and then access the certificates using \"/proc/<proc number>/root/<file-path>\". Pass these certificate file paths to the etcdctl command :)"
```
```suggestion
  echo "You can now start using etcdctl. Just execute \"etcdctl\" to use it. See https://etcd.io/docs/v3.4/dev-guide/interacting_v3/ for more details."
  echo "This tool assumes that it is being run in an ephemeral container in a pod. If this is not the case, please ensure that you provide the correct certificates, the correct endpoint, and have all necessary accesses."
  echo "Certificates to be passed to the command should be mounted onto any container in the pod having a shared process namespace. Please run \"ps -A\" to list all processes and then access the certificates using \"/proc/<proc number>/root/<file-path>\". Pass these certificate file paths to the etcdctl command :)"
```

petersutter commented:
Why don't we just install the latest version instead of asking the user?

mimiteto commented:
I just took two current scripts - `dotfiles/.install_on_demand/.wireguard` and `hacks/install_etcdctl`.
In the hacks script there is an option for the version, that's why it's transformed to a question (that's how it's handled in the other scripts).
Just a note, after discussing with @plkokanov - current PR will not be actually that useful.
If you decide to use the `ops-toolbelt` with non-root user you actually don't have access to dotfiles, so additional change is needed (for example, loading dotfiles in `/etc/bashrc`).
I still haven't gotten around to check this approach.

petersutter commented:
> In the hacks script there is an option for the version, that's why it's transformed to a question (that's how it's handled in the other scripts).

I think we should just install the latest version, if the user runs `etcdctl` or `wg`. If they want to have a specific version, they should run the install script and pass the version as argument

petersutter commented:
btw. I have also created https://github.com/gardener/ops-toolbelt/issues/104 for k9s as we have an install script for this as well but does not yet support to install it on demand

mimiteto commented:
```
If they want to have a specific version, they should run the install script and pass the version as argument
```
Arguments are passed directly to the binaries after they are installed, so I am not sure if that's a good idea.

IMO having the option to specify version can come in handy and first invocation of those commands will anyway break whatever script (because of the y/n part) so I don't see a reason to currently remove versions.

It would make sense if we decide to change it so when an operator invokes, for example, `wg` if it's not installed to automatically install, no questions asked. In that case I agree having the option to specify a version is pointless.

petersutter commented:
What about the following:
- The install script should not invoke the binary. It should just install it. If you pass a version as argument it should take it, otherwise it should install the latest (or maybe a hardcoded version, depends on how we want it to behave).
- Running the binary, like `etcdctl` should install it on demand in case it is not already installed, no questions asked to the user. We could redirect any output of the install script to `stderr`, which could make it more compatible with scripts calling it for the first time..

mimiteto commented on hacks/install_etcdctl:
```diff
@@ -21,27 +21,36 @@ function show_help () {
 
 function install () {
-  local version=$1
+  local version="${1:-v3.4.30}" # default version "v3.4.3
```
I was thinking about setting the default version to the latest from `etcd` github release, but I saw that they are tagging latest released as `latest` and not the highest version (version `3.4.xx` may get tagged as higher than `3.5.xx`)

petersutter replied:
we could let [renovate](https://github.com/gardener/ops-toolbelt/pull/107) manage the version. I just tested it and it took `v3.5.12` instead of `v3.4.31`, which is currently tagged as latest
![Screenshot 2024-03-25 at 09 52 13](https://github.com/gardener/ops-toolbelt/assets/5526658/0ead7f58-765c-49de-bbbb-d069a9b30ee5)

mimiteto replied:
Should I adopt the renovate on boarding PR as well?

petersutter replied:
no, either this PR gets merged first THEN the onboarding PR can be adapted, or the onboarding PR gets merged first then we can adapt this PR

mimiteto commented:
I made the on-demand binaries get installed when executed without any questions asked.
I also changed the `etcdctl` install script to consider current user (root goes to the system path, non-root uses dir relevant to the current user).
In addition I made the `etcdctl` alias add arguments about certificates in the default paths when `etcdctl` is executed within debug pod.
PR is currently incomplete, as it still needs to change the way `dotfiles` are made available

petersutter commented on hacks/install_etcdctl:
```diff
@@ -21,27 +21,36 @@ function show_help () {
-    fi
+    dest="${DOTFILES_HOME}/bin/etcdctl"
   fi
+  arch="$(arch)"
```
If I run `arch` in the container, it returns `x86_64` and hence the download url is wrong

petersutter commented:
> PR is currently incomplete

do you want to change it to `draft` then?

petersutter commented on hacks/install_etcdctl:
```diff
@@ -21,27 +21,42 @@ function show_help () {
 
+  curl -sL "${download_url}" \
+    -o "${pkg}.tar.gz" && \
+    tar -zxvf "${pkg}.tar.gz" && \
```
```suggestion
    tar -zxf "${pkg}.tar.gz" && \
```
too much noise otherwise..

mimiteto replied:
TBH in the case of the `apt` installed packages I would not care for that output that much - `apt` will try it's best to install and if it's not able to do so IRL you will not want to debug the toolbelt image while you are using it. At best you would try installing it again and then shift to the important task at hand.
On the other hand - we are installing `etcdctl` (and `k9s`) with simple `curl` commands and naively unpacking it. A lot of things can go south and having this output may make things simple.
My assumption is that you are using the toolbelt in extraordinary situation.

petersutter replied:
Currently, the info message that we print at the bottom gets lost among all the outputs from the `tar zxvf` command.
https://github.com/gardener/ops-toolbelt/blob/aa126683ab61c73bc6851a52e6ee24be7289ecc7/hacks/install_etcdctl#L58-L60

If you do not want to drop the verbose option, at least we should separate the info message with an additional newline from the extraction output.

mimiteto replied:
Colored it in yellow.
Added same for k9s.

mimiteto commented:
We do print it for k9s.
I don't mind the message and also don't think it's that necessary.
Ref: https://github.com/gardener/ops-toolbelt/blob/aa126683ab61c73bc6851a52e6ee24be7289ecc7/hacks/install_k9s#L38

petersutter commented:
what I wanted to say is that it's currently inconsistent. Either we always print the install info message or not. But we can also postpone this and align it in a future PR

mimiteto commented:
@plkokanov @petersutter I need your opinion - in order to make the `etcdctl` install accessible/usable from all users, I did the following:
* Added a new dir with open permissions [ref](https://github.com/gardener/ops-toolbelt/pull/102/files#diff-585e09338be8d61489c23232e5fe922dae7a54726fcbc6ae95e242750ecdbdcaR116-R119)
* Moved the on-demand installation to `/etc/bash` - [ref](https://github.com/gardener/ops-toolbelt/pull/102/files#diff-585e09338be8d61489c23232e5fe922dae7a54726fcbc6ae95e242750ecdbdcaR116-R119)
* Added the new `opt` dir to the path of all users - [ref](https://github.com/gardener/ops-toolbelt/pull/102/files#diff-585e09338be8d61489c23232e5fe922dae7a54726fcbc6ae95e242750ecdbdcaR157-R161)

The alternative option I see is to populate the `/nonroot/hacks` directory and install the binary there. I didn't chose that approach, as I am not aware if it's a good idea to make that directory with wide permissions.

WDYT?

petersutter commented on hacks/install_etcdctl:
```diff
@@ -21,28 +21,54 @@ function show_help () {
 function install () {
-  local version=$1
+  # renovate: datasource=github-releases depName=etcd-io/etcd
+  _version="v3.5.12"
```
name it `etcd_version`, or we should adapt the regex so that it also accepts just `version`

plkokanov replied:
I presonally prefer `etcd_version`

petersutter commented:
> The alternative option I see is to populate the `/nonroot/hacks` directory and install the binary there. I didn't chose that approach, as I am not aware if it's a good idea to make that directory with wide permissions.

I think the `/opt/bin` solution is fine

> Also, should I add this set of configs on the node when running the [ops pod](https://github.com/gardener/ops-toolbelt/blob/master/hacks/ops-pod#L186-L191)
> 
> WDYT?

This would make it consistently available right? Do not see a reason that would speak against it..

ashwani2k commented on hacks/install_etcdctl:
```diff
@@ -12,28 +12,54 @@ function show_help () {
 function install () {
-  local version=$1
+  # renovate: datasource=github-releases depName=etcd-io/etcd
+  etcd_version="v3.5.12"
```
Shall we make the default version to be the one used by Gardener internally i.e. 3.4.26 and if someone wants to override they can do so they can pass in the argument. 

Otherwise for any gardener operator he needs to always do an additional step of running with 3.4.26 which defeats the original purpose.

mimiteto replied:
That's a good point.
With the current config we will always get a PR for the latest because of renovate, so if changed we will get PR for the latest anyway.
Where should we read the version, so it's not maintained manually?
Or should we leave it and keep it up manually?

ashwani2k replied:
Currently we don't upgrade etcd version so frequently, so I'll just keep it manual and when we move to 3.5.x this year as part of this https://github.com/gardener/etcd-druid/issues/445 then we will also adopt the ops-toolbelt accordingly. wdyt?

petersutter replied:
> Otherwise for any gardener operator he needs to always do an additional step of running with 3.4.26 which defeats the original purpose.

ok I guess in this case we should hardcode it and not let it be managed by renovate to update to the latest version. 
Alternatively, @ashwani2k, if the patch version does not matter, we could configure renovate to only update the patch version, like we do it for `kubectl` https://github.com/gardener/ops-toolbelt/blob/13186b58b0bcb0dd183c18f1de892311ff60a95b/.github/renovate.json5#L46

mimiteto replied:
I've set it to a hardcoded version.

ashwani2k replied:
No the patch version matters as we had to upgrade to 3.4.26 from 3.4.13 as it fixed some of the issues which would have forced us to migrate it to 3.5.x instead. 
But i'll ask the etcd colleagues to confirm this, if it really matters or just 3.4 is enough. 
cc @shreyas-s-rao @ishan16696 @aaronfern

shreyas-s-rao replied:
Generally the patch version doesn't matter from an API perspective. But it's generally recommended to use the same etcdctl version as the etcd that it is used to interact with, since they are well-tested together. In the case of gardener currently, we use v3.4.26 for both the etcd and etcdctl, so we can retain that.

mimiteto replied:
As there are no new comments and static version is set, I will resolve this conversation.

petersutter commented on hacks/ops-pod:
```diff
@@ -186,9 +186,14 @@ if [[ ${node_chroot} -eq ${TRUE} ]]; then
 else
   kubectl -n $namespace exec -ti $name -- bash -c "echo -e '\nNode root dir is mounted under /host' >> /etc/motd; /bin/bash"
-fi
+i
```
```suggestion
fi
```
coud you also fix the default image to be

```
default_image="[europe-docker.pkg.dev/gardener-project/releases/gardener/ops-toolbelt:latest](http://europe-docker.pkg.dev/gardener-project/releases/gardener/ops-toolbelt:latest)"
```

mimiteto commented:
After a sync with @petersutter:
* I've reverted the `ops-pod` script
* I've dropped the image pull policy - that way we will default to `Always` if no image is specified, because default image in the script has `latest` tag. Otherwise we will default to the `IfNotPresent`

petersutter reviewed (approved):
/lgtm

plkokanov commented on hacks/install_k9s:
```diff
@@ -14,26 +14,53 @@ function show_help () {
+  local pkg
+  local dest
+  local yellow="\033[0;33m"
+  local nc="\033[0m"
```
The `yellow` and `nc` vars appear twice.

mimiteto replied:
No longer like that.

plkokanov commented on hacks/start_chrooted:
```diff
@@ -0,0 +1,33 @@
+
+trap cleanup EXIT
+
+if [ "${#}" -eq 0 ] || [ "${1}" -ne "1" ]; then
```
Do we need this check in a file called `start_chrooted`. I find it a bit confusing.

mimiteto replied:
File needs a better name (possibly `spawn-ops-pod-shell`).

plkokanov commented on hacks/ops-pod:
```diff
@@ -183,12 +182,12 @@ while [[ $(kubectl -n $namespace get pods | sed -n -r "s/^$name.*Running.*$/Runn
 # exec into pod (and chroot into node if a node was selected)
 if [[ ${node_chroot} -eq ${TRUE} ]]; then
-  kubectl -n $namespace exec -ti $name -- bash -c "rm -rf /host/root/dotfiles 1> /dev/null; \
+  kubectl -n $namespace exec -ti $name -- bash -c 'rm -rf /host/root/dotfiles 1> /dev/null; \
```
Why don't we use the `/hacks/start_chrooted` script here?

mimiteto replied:
We synced with @petersutter .
We should leave the setup and clean-up for the on-demand scripts for another PR.
I will remove the `start_chrooted` as well.

plkokanov commented on hacks/install_etcdctl:
```diff
@@ -12,30 +12,47 @@ function show_help () {
 
+  curl -sL "${download_url}" \
+    -o "${tmp_dir}/${pkg}.tar.gz" && \
+    tar -zxvf "${tmp_dir}/${pkg}.tar.gz" -C "${tmp_dir}/dest" && \
```
Can we remove the `v` option from `tar`? We could probably replace it with an `echo`s that says that we are downloading it and then an echo that says that we are installing/unarchiving the app.
Same for the other tools provided via `install_on_demand`.

mimiteto replied:
Seems like the popular opinion. I will just remove the verbose parts.

mimiteto replied:
Verbose options are now removed for both `etcdctl` and `k9s`.

plkokanov reviewed (changes requested):
Just a few more minor suggestions. Rest looks very good, I think.

plkokanov commented on install_on_demand/.etcdctl:
```diff
@@ -0,0 +1,76 @@
+#! /usr/bin/env bash
+
```
Add license header.

plkokanov commented on install_on_demand/.k9s:
```diff
@@ -0,0 +1,18 @@
+#! /usr/bin/env bash
+
```
Add license header.

plkokanov commented on install_on_demand/.shrc:
```diff
@@ -0,0 +1,6 @@
+source /etc/bash/install_on_demand/.table
```
Add license header

plkokanov commented on install_on_demand/.table:
```diff
@@ -1,19 +1,10 @@
+#! /usr/bin/env bash
+
```
Add license header.

plkokanov commented on install_on_demand/.wireguard:
```diff
@@ -0,0 +1,9 @@
+#! /usr/bin/env bash
+
```
Add license header.
